File: backend/src/main/java/com/technicalescaperoom/backend/controller/admin/AdminResultsController.java
package com.technicalescaperoom.backend.controller.admin;

import com.technicalescaperoom.backend.config.security.AdminPrincipal;
import com.technicalescaperoom.backend.dto.admin.EventStatisticsDto;
import com.technicalescaperoom.backend.dto.admin.LeaderboardEntryDto;
import com.technicalescaperoom.backend.service.admin.LeaderboardService;
import com.technicalescaperoom.backend.service.admin.ResultExportService;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpHeaders;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("/api/admin/events/{eventId}")
@RequiredArgsConstructor
public class AdminResultsController {

    private final LeaderboardService leaderboardService;
    private final ResultExportService resultExportService;

    @GetMapping("/leaderboard")
    public ResponseEntity<List<LeaderboardEntryDto>> getLeaderboard(@PathVariable Long eventId) {
        List<LeaderboardEntryDto> leaderboard = leaderboardService.getLeaderboard(eventId);
        return ResponseEntity.ok(leaderboard);
    }

    @GetMapping("/statistics")
    public ResponseEntity<EventStatisticsDto> getEventStatistics(@PathVariable Long eventId) {
        EventStatisticsDto stats = leaderboardService.getEventStatistics(eventId);
        return ResponseEntity.ok(stats);
    }

    @GetMapping(value = "/export/results", produces = "text/csv")
    public ResponseEntity<String> exportResultsCsv(
            @PathVariable Long eventId,
            @AuthenticationPrincipal AdminPrincipal principal
    ) {
        String csv = resultExportService.generateEventResultsCsv(principal, eventId);
        return ResponseEntity.ok()
                .header(HttpHeaders.CONTENT_DISPOSITION, "attachment; filename=\"codexcape_event_" + eventId + "_results.csv\"")
                .contentType(MediaType.parseMediaType("text/csv"))
                .body(csv);
    }

    @GetMapping(value = "/export/progress", produces = "text/csv")
    public ResponseEntity<String> exportProgressCsv(
            @PathVariable Long eventId,
            @AuthenticationPrincipal AdminPrincipal principal
    ) {
        String csv = resultExportService.generateTeamProgressCsv(principal, eventId);
        return ResponseEntity.ok()
                .header(HttpHeaders.CONTENT_DISPOSITION, "attachment; filename=\"codexcape_event_" + eventId + "_progress.csv\"")
                .contentType(MediaType.parseMediaType("text/csv"))
                .body(csv);
    }
}

File: backend/controllers/admin/adminResultsController.js
const leaderboardService = require('../../services/admin/leaderboardService');
const resultExportService = require('../../services/admin/resultExportService');
const scoringService = require('../../services/scoringService');

class AdminResultsController {
  async getLeaderboard(req, res, next) {
    try {
      const eventId = parseInt(req.params.eventId, 10);
      const leaderboard = await leaderboardService.getLeaderboard(eventId);
      res.json(leaderboard);
    } catch (err) {
      next(err);
    }
  }

  async getTeamScoreSummary(req, res, next) {
    try {
      const teamId = parseInt(req.params.teamId, 10);
      const summary = await scoringService.getTeamScoreSummary(teamId);
      res.json(summary);
    } catch (err) {
      next(err);
    }
  }

  async getTeamScoreEvents(req, res, next) {
    try {
      const teamId = parseInt(req.params.teamId, 10);
      const events = await scoringService.getTeamScoreEvents(teamId);
      res.json(events);
    } catch (err) {
      next(err);
    }
  }

  async getEventStatistics(req, res, next) {
    try {
      const eventId = parseInt(req.params.eventId, 10);
      const stats = await leaderboardService.getEventStatistics(eventId);
      res.json(stats);
    } catch (err) {
      next(err);
    }
  }

  async exportResultsCsv(req, res, next) {
    try {
      const eventId = parseInt(req.params.eventId, 10);
      const csv = await resultExportService.generateEventResultsCsv(req.admin, eventId);
      res.setHeader('Content-Disposition', `attachment; filename="codexcape_event_${eventId}_results.csv"`);
      res.setHeader('Content-Type', 'text/csv');
      res.send(csv);
    } catch (err) {
      next(err);
    }
  }

  async exportProgressCsv(req, res, next) {
    try {
      const eventId = parseInt(req.params.eventId, 10);
      const csv = await resultExportService.generateTeamProgressCsv(req.admin, eventId);
      res.setHeader('Content-Disposition', `attachment; filename="codexcape_event_${eventId}_progress.csv"`);
      res.setHeader('Content-Type', 'text/csv');
      res.send(csv);
    } catch (err) {
      next(err);
    }
  }
}

module.exports = new AdminResultsController();
